pkg/cosign/kubernetes/webhook/validator.go

```diff
@@ -166,24 +216,106 @@ func (v *Validator) validatePodSpec(ctx context.Context, ps *corev1.PodSpec, opt
+// the image was not validated against any matching policy and hence authority.
+func validatePolicies(ctx context.Context, ref name.Reference, defaultKC authn.Keychain, policies map[string][]v1alpha1.Authority, _ ...ociremote.Option) (int, map[string][]error) {
+	// For a policy that does not pass at least one authority, gather errors
+	// here so that we can give meaningful errors to the user.
+	ret := map[string][]error{}
+	// For each matching policy it must validate at least one Authority within
+	// it.
+	// From the Design document, the part about multiple Policies matching:
+	// "If multiple policies match a particular image, then ALL of those
+	// policies must be satisfied for the image to be admitted."
+	// So we keep a tally to make sure that all the policies matched.
+	policiesValidated := 0
+	// If none of the Authorities for a given policy pass the checks, gather
+	// the errors here. If one passes, do not return the errors.
+	authorityErrors := []error{}
+	for p, authorities := range policies {
+		logging.FromContext(ctx).Debugf("Checking Policy: %s", p)
+		// Now the part about having multiple Authority sections within a
+		// policy, any single match will do:
+		// "When multiple authorities are specified, any of them may be used
+		// to source the valid signature we are looking for to admit an image.""
+		authoritiesValidated := 0
+		for _, authority := range authorities {
+			logging.FromContext(ctx).Debugf("Checking Authority: %+v", authority)
+			// TODO(vaikas): We currently only use the defaultKC, we have to look
+			// at authority.Sources to determine additional information for the
+			// WithRemoteOptions below, at least the 'TargetRepository'
+			// https://github.com/sigstore/cosign/issues/1651
+			opts := ociremote.WithRemoteOptions(remote.WithAuthFromKeychain(defaultKC))
 
-	for _, authority := range authorities {
-		if authority.Key != nil {
-			// Get the key from authority data
-			if authorityKeys, fieldErr := parseAuthorityKeys(ctx, authority.Key.Data); fieldErr != nil {
-				errs = errs.Also(fieldErr)
-			} else {
-				keys = append(keys, authorityKeys...)
+			if authority.Key != nil {
+				// Get the key from authority data
+				if authorityKeys, fieldErr := parseAuthorityKeys(ctx, authority.Key.Data); fieldErr != nil {
+					authorityErrors = append(authorityErrors, errors.Wrap(fieldErr, "failed to parse Key values"))
+				} else {
+					// TODO(vaikas): What should happen if there are multiple keys
+					// Is it even allowed? 'valid' returns success if any key
+					// matches.
+					// https://github.com/sigstore/cosign/issues/1652
+					if err := valid(ctx, ref, authorityKeys, opts); err != nil {
+						authorityErrors = append(authorityErrors, errors.Wrap(err, "failed to validate keys"))
+						continue
+					}
+					// This authority matched, so mark it as validated and
+					// continue through other policies, no need to look at more
+					// of the Authorities.
+					authoritiesValidated++
```

Comment:
Perhaps instead of just a count, we can have this return the valid signatures instead?

I think this is what `cosign.VerifySignatures` does (something like: `(validSigs []oci.Signature, bundleVerified bool, err error)`